rc/lsp.kak

```diff
@@ -2123,18 +2153,15 @@ decl -hidden int-list lsp_snippets_placeholder_groups
 face global SnippetsNextPlaceholders black,green+F
 face global SnippetsOtherPlaceholders black,yellow+F
 
-# First param is the text that was inserted in the completion, which will be deleted
-# Second param is the actual snippet
-def -hidden lsp-snippets-insert-completion -params 2 %{ eval -save-regs "a" %{
-    reg 'a' "%arg{1}"
-    exec -draft "<a-;><a-/><c-r>a<ret>d"
-    eval -draft -verbatim lsp-snippets-insert %arg{2}
-    remove-hooks window lsp-post-completion
-    hook -once -group lsp-post-completion window InsertCompletionHide .* %{
-        try %{
-            lsp-snippets-select-next-placeholders
-            exec '<a-;>d'
-        }
+declare-option -hidden str lsp_snippet_to_insert ""
+def -hidden lsp-snippets-insert-completion -params 1 %{ eval %{
+    set window lsp_snippet_to_insert %arg{1}
+    lsp-completion-on-dismiss %{
+        # Delete the inserted text
+        select %opt{lsp_completion_selection}
+        exec '<a-;>d'
```

Comment:
Wow Kendrick is back! I've almost given up hope, same for proper snippets support. I tried to write a similar patch last week but quickly gave up.
Your approach looks good, I'll take a closer look tomorrow.

There seems to be an issue still with gopls: the right parenthesis of a call is deleted.
I'm sure we can fix that one way or another. If worst comes to worst we can even make a temporary RawKey hook.
But since that was already broken the same way it's not a blocker.

>  - Snippets are now only fully expanded when the completions menu is closed, making them much more robust

yep, given how Kakoune's completion engine works that's definitely the way to go

> This could be used to implement #40 pretty easily,

Right. I think we already support most text edits you'd find in the wild, let me know if you find a missing one. Then we have motivation to add full support.

Reply:
> There seems to be an issue still with gopls: the right parenthesis of a call is deleted.

Hmm, i thought i fixed this on the last force push by replacing empty placeholders with a single space, but i may have missed some cases.
Placeholder parsing and jumping should probably be fully rewritten though. Maybe using replace-ranges to show the placeholder instead, though this also has its issues.

Reply:
Here is a failing test that shows the problem with gopls. Anyway, we can fix this later

```diff
diff --git a/test/gopls-completion-snippet.sh b/test/gopls-completion-snippet.sh
new file mode 100755
index 00000000..0ca60e9c
--- /dev/null
+++ b/test/gopls-completion-snippet.sh
@@ -0,0 +1,34 @@
+#!/bin/sh
+
+# REQUIRES: command -v gopls
+
+. test/lib.sh
+
+cat > .config/kak-lsp/kak-lsp.toml << EOF
+snippet_support = true
+[language.go]
+filetypes = ["go"]
+roots = ["*.go"]
+command = "gopls"
+EOF
+
+
+cat > main.go << EOF
+package main
+
+import "os"
+
+func foo(i int, j int) {
+}
+EOF
+
+test_tmux_kak_start 'edit main.go'
+
+test_tmux send-keys 4jo fo
+test_sleep
+test_tmux send-keys C-n
+test_sleep
+test_tmux send-keys 1
+test_sleep
+test_tmux capture-pane -p | grep '^\s*foo'
+# CHECK: foo(1)
```

Reply:
OK we were just missing a case for gopls:

```diff
diff --git a/rc/lsp.kak b/rc/lsp.kak
index 76c69a49..565d11ac 100644
--- a/rc/lsp.kak
+++ b/rc/lsp.kak
@@ -2261,6 +2275,8 @@ foreach my $i (0 .. $#sel_content) {
         $def =~ s/\}\}/}/g;
         # double up single-quotes
         $def =~ s/'\''/'\'''\''/g;
+        # avoid empty placeholders because they complicate movement
+        if (length $def == 0) { $def = " " };
         print(" '\''$def'\''");
     } else {
         print(" '\'' '\''");
```

Reply:
Ah, that makes a lot of sense!